### ISSUE TITLE
Can't authenticate when specifying domain name for project and user

### ISSUE THREAD

**Issue by oblutak** [flystack/misty#115]
Just discovered misty gem but can't seem to authenticate properly to private openstack cloud.  I think the issue can be summarized to: what authentication hash to I need to provide to make the following test succeed:
```ruby
it 'authenticates using a domain name with username and project name' do
  auth = { }

  stub_request(:post, 'http://localhost:5000/v3/auth/tokens').
    with(:body => %q`{"auth":{"identity":{"methods":["password"],"password":{"user":{"name":"user_name","domain":{"name":"domain_name"},"password":"secret"}}},"scope":{"project":{"name":"project_name","domain":{"name":"domain_name"}}}}}`).
    to_return(:status => 200, :body => JSON.dump(auth_response_v3('identity', 'keystone')), :headers => {'x-subject-token'=>'token_data'})

  Misty::AuthV3.new(auth, config)
end
```

**Comment by oblutak** [comment]
I thought the following should have worked
```ruby
auth = {
  :url      => 'http://localhost:5000',
  :user  => 'user_name',
  :password => 'secret',
  :user_domain => 'domain_name',
  :project  => 'project_name',
  :project_domain => 'domain_name'
}
```
but my first attempt was:
```ruby
auth = {
  :url      => 'http://localhost:5000',
  :user  => 'user_name',
  :password => 'secret',
  :domain => 'domain_name',
  :project  => 'project_name',
}
```
I am fairly new to openstack.  I have been using vagrant with the openstack provider, with some success, but need more flexibility and was looking a misty for that.  The test above was derived from what I am seeing when using vagrant.

**Comment by gildub** [comment]
@oblutak, with Identity (Keystone) version 3 the important part is effectively the domain, if no domain is provided in the authentication credentials then Misty will use the default domain (its name is 'Default' and its ID is 'default').  Meanwhile depending on your OpenStack set-up the default domain might be different.

It seems your first attempt didn't work because the project name is not enough for Keystone to identify without ambiguity the scope (project names are not necessarily unique). Please try by providing either a project_id or a project_domain.

Also the credentials you might provide to Openstack Client (CLI) should also work.

**Comment by oblutak** [comment]
In the first example, I tried to explicitly set a project_domain (and a project) as well as a user_domain.  What I am seeing is that project and project_domain are used, but the 'user_domain' is set back to the default domain_id of 'default'.  It appears I need to use a non-default user_domain on my openstack.  I am not sure if it is always the case, but project_domain == user_domain, which is not equal to 'default'.

**Comment by gildub** [comment]
@oblutak, is this issue resolved? If not is your cloud accessible remotely so I can help?

**Comment by oblutak** [comment]
@gildub, I don’t consider it resolved.  I won’t have access to a terminal until Monday. I will follow up early next week. Unfortunately it is a private cloud and I have no way to give remote access. 
One thing I am confuse is the support/intent of having support for:
Domain, user_domain and project_domain

**Comment by gildub** [comment]
@oblutak, 

No worries regarding access to the private cloud, we'll have to debug/trace if needed. Meanwhile before that what would help is to confirm the authentication works with your credentials by obtaining a token, and if not, help identifying the missing pieces. 

For more information about Keystone version 3 and the domain name space:
* https://prosuncsedu.wordpress.com/2016/03/09/relationship-among-concepts-domain-project-role-user-group-user-and-token-in-openstack-keystone/
* https://www.mirantis.com/blog/manage-openstack-projects-using-domains-havana/
* https://adam.younglogic.com/2016/11/keystone-domains-are-projects/

Basically you should be able to do something like the following, using CURL.

```shell
export OS_AUTH_URL="http://192.168.0.1:5000/v3/auth/tokens"

export OS_USER=admin
export OS_PASSWORD="secret"
export OS_USER_DOMAIN_ID=default

export OS_PROJECT_DOMAIN_ID=default
export OS_PROJECT_NAME=admin

# Note this won't work if there space in any variable (more likely the password), use hard coded value in -d section (or more ticks)

curl -i \
  -H "Content-Type: application/json" \
  -d '
{ "auth": {
    "identity": {
      "methods": ["password"],
      "password": {
        "user": {
          "name": "'"${OS_USER}"'",
          "domain": { "id": "'"${OS_USER_DOMAIN_ID}"'" },
          "password": "'"${OS_PASSWORD}"'"
        }
      }
    },
    "scope": {
      "project": {
        "name": "'"${OS_PROJECT_NAME}"'",
        "domain": { "id": "'"${OS_PROJECT_DOMAIN_ID}"'" }
      }
    }
  }
}' \
  ${OS_AUTH_URL} ; echo
```

If that works then you're authenticated, scroll back to the head section for the token:
`X-Subject-Token: eNSSSLWamf2dzEDaLEcRmqprIG76-yUgxN4EwF77X2-kQEnj1MMRlSiTLrwi6BDOIgwcvQVSttapcD6PsILfXSyzN004HellaooD65hVfclHOQ3Y47_W-1O
7TzH1XEM33Pq6mXBENOK1kvjSu7Sm5l10DeoijdlnpXAxl_AQ9bLECktxe3PiXs`

Then you should be able to use this domain scoped token to perform requests:
```shell
export OS_TOKEN="eNSSSLWamf2dzEDaLEcRmqprIG76-yUgxN4EwF77X2-kQEnj1MMRlSiTLrwi6BDOIgwcvQVSttapcD6PsILfXSyzN004HellaooD65hVfclHOQ3Y47_W-1O
7TzH1XEM33Pq6mXBENOK1kvjSu7Sm5l10DeoijdlnpXAxl_AQ9bLECktxe3PiXs"

curl -s -H "X-Auth-Token: $OS_TOKEN" http://192.168.0.1/9696/v2.0/networks | python -mjson.tool
```

Let's follow up early next week.

**Comment by oblutak** [comment]
@gildub 

Thanks for the link.  I didn't read all of it yet.

I can get the above curl to work, but I want to be able to use domain name and not domain id.  The private cloud in use is not using the 'default' domain, and getting the domain id for a non-default domain is not trivial for a regular user.  I think the test in my initial comment properly express what I am looking/expecting from misty, which expressed using curl would map to

```bash
export OS_AUTH_URL="http://192.168.0.1:5000/v3/auth/tokens"

export OS_USER=admin
export OS_PASSWORD="secret"
export OS_USER_DOMAIN_NAME=CustomDomain

export OS_PROJECT_DOMAIN_NAME=CustomDomain
export OS_PROJECT_NAME=admin

# Note this won't work if there space in any variable (more likely the password), use hard coded value in -d section (or more ticks)

curl -i \
  -H "Content-Type: application/json" \
  -d '
{ "auth": {
    "identity": {
      "methods": ["password"],
      "password": {
        "user": {
          "name": "'"${OS_USER}"'",
          "domain": { "name": "'"${OS_USER_DOMAIN_NAME}"'" },
          "password": "'"${OS_PASSWORD}"'"
        }
      }
    },
    "scope": {
      "project": {
        "name": "'"${OS_PROJECT_NAME}"'",
        "domain": { "name": "'"${OS_PROJECT_DOMAIN_NAME}"'" }
      }
    }
  }
}' \
  ${OS_AUTH_URL} ; echo
```

**Comment by gildub** [comment]
@oblutak, just to be clear, the previous curl is working for you and let you obtain a token, then the following should work using only domain names:

```ruby
auth = {
  :url  => 'http://localhost:5000',
  :user => 'user_name',
  :user_domain => 'domain_name',
  :password => 'secret',
  :project  => 'project_name',
  :project_domain => 'domain_name'
}
```

**Comment by oblutak** [comment]
@gildub , your previous comment is correct.  The curl works on my system, but the auth hash provided does not work.  Sorry it took so long to get here, but that is what I was trying to express with the sample unit test.  With the auth hash provided, the test fails because it looks like misty is ignoring my user_domain setting and reverts back to using the 'default' domain id.
```
registered request stubs:

stub_request(:post, "http://localhost:5000/v3/auth/tokens").
  with(:body => "{\"auth\":{\"identity\":{\"methods\":[\"password\"],\"password\":{\"user\":{\"name\":\"user_name\",\"domain\":{\"name\":\"domain_name\"},\"password\":\"secret\"}}},\"scope\":{\"project\":{\"name\":\"project_name\",\"domain\":{\"name\":\"domain_name\"}}}}}")

Body diff:
 [["-", "auth.identity.password.user.domain.id", "default"],
 ["+", "auth.identity.password.user.domain.name", "domain_name"]]
```

**Comment by gildub** [comment]
@oblutak, no worries, first off, thank you for bringing this up. Also I'm sorry it took me longer than needed and thank your for your patience.
Effectively, I see you point. It seems I'm not facing the issue because in my tests I'm assuming the default domain id is set as safe net. 

I'm going to investigate and fix this accordingly.

**Comment by gildub** [comment]
The fix is now available in [v1.4.0](https://github.com/flystack/misty/releases/tag/v1.4.0). Please let me know if that works for you. Cheers

**Comment by oblutak** [comment]
It works!  Thanks for the quick fix and putting out a release.  Also thanks for your patience and providing the links to educate me a bit more.  Much to learn on Openstack...  Now time to play with misty!
